docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,259 @@
+   via a custom socket protocol that sends requests and responses over a
+   Unix-domain or TCP socket connection as length-prefixed protocol buffers.
+   In Tendermint, this is handled by the [socket client][socket-client].
+
+3. **gRPC**: Out-of-process applications may export the ABCI service via gRPC.
+   In Tendermint, this is handled by the [gRPC client][grpc-client].
+
+Both the out-of-process options (2) and (3) have a long history in Tendermint.
+The beginnings of the gRPC client were added in [May 2016][abci-start] when
+ABCI was still hosted in a separate repository, and the socket client (formerly
+called the "remote client") was part of ABCI from its inception in November
+2015.
+
+At that time when ABCI was first being developed, the gRPC project was very new
+(it launched Q4 2015) and it was not an obvious choice for use in Tendermint.
+It took a while before the language coverage and quality of gRPC reached a
+point where it could be a viable solution for out-of-process applications.  For
+that reason, it made sense for the initial design of ABCI to focus on a custom
+protocol for out-of-process applications.
+
+## Problem Statement
+
+For practical reasons, ABCI needs an interprocess communication option to
+support applications not written in Go. The two practical options are RPC and
+FFI, and for operational reasons an RPC mechanism makes more sense.
+
+The socket protocol has not changed all that substantially since its original
+design, and has the advantage of being simple to implement in almost any
+reasonable language.  However, its simplicity includes some limitations that
+have had a negative impact on the stability and performance of out-of-process
+applications using it. In particular:
+
+- The protocol lacks request identifiers, so the client and server must return
+  responses in strict FIFO order. Even if the client issues requests that have
+  no dependency on each other, the protocol has no way except order of issue to
+  map responses to requests.
+
+  This reduces (in some cases substantially) the concurrency an application can
+  exploit, since the parallelism of requests in flight is gated by the slowest
+  active request at any moment.  There have been complaints from some network
+  operators on that basis.
+
+- The protocol lacks method identifiers, so the only way for the client and
+  server to understand which operation is requested is to dispatch on the type
+  of the request and response payloads. For responses, this means that [any
+  error condition is terminal not only to the request, but to the entire ABCI
+  client](https://github.com/tendermint/tendermint/blob/master/abci/client/socket_client.go#L149).
+
+  The historical intent of terminating for any error seems to have been that
+  all ABCI errors are unrecoverable and hence protocol fatal <!-- markdown-link-check-disable-next-line -->
+  (see [Note 1](#note1)).  In practice, however, this greatly complicates
```

Comment:
This note is super useful, and would def be good to distinguish between mechanical and protocol errors, but do we have examples of trouble with mechanical errors or of challenges debugging?